logbook-ktor-server/src/main/kotlin/org/zalando/logbook/server/LogbookServer.kt

```diff
@@ -32,25 +29,37 @@ class LogbookServer(
     companion object : BaseApplicationPlugin<Application, Config, LogbookServer> {
         private val responseProcessingStageKey: AttributeKey<Logbook.ResponseProcessingStage> =
             AttributeKey("Logbook.ResponseProcessingStage")
+        private val proceedWithKey: AttributeKey<ByteReadChannel> =
+            AttributeKey("Logbook.proceedWith")
+
         override val key: AttributeKey<LogbookServer> = AttributeKey("LogbookServer")
+
         override fun install(pipeline: Application, configure: Config.() -> Unit): LogbookServer {
             val config = Config().apply(configure)
             val plugin = LogbookServer(config.logbook)
 
-            pipeline.receivePipeline.intercept(ApplicationReceivePipeline.Before) {
+            // This interceptor is executed before ApplicationReceivePipeline,
+            // and regardless of whether the request has body or not
+            pipeline.intercept(ApplicationCallPipeline.Call) {
                 val request = ServerRequest(call.request)
                 val requestWritingStage = plugin.logbook.process(request)
-                val proceedWith = when {
-                    request.shouldBuffer() && !call.request.receiveChannel().isClosedForRead -> {
-                        val content = call.request.receiveChannel().readBytes()
-                        request.buffer(content)
-                        ByteReadChannel(content)
-                    }
 
-                    else -> it
+                val proceedWith = if (!call.request.receiveChannel().isClosedForRead) {
+                    val content = call.request.receiveChannel().readBytes()
+                    request.buffer(content)
+                    ByteReadChannel(content)
+                } else {
+                    ByteReadChannel.Empty
                 }
+
                 val responseProcessingStage = requestWritingStage.write()
                 call.attributes.put(responseProcessingStageKey, responseProcessingStage)
+                call.attributes.put(proceedWithKey, proceedWith)
+            }
+
+            // This interceptor is executed only when the request has body
+            pipeline.receivePipeline.intercept(ApplicationReceivePipeline.Transform) { subject ->
```

Comment:
Similarly to my comment on the attribute, do we actually need this pipeline interceptor ?

Reply:
Same as above :))